Liquid-solid fluidized bed example (#845)
lethe-cfd/lethe: doc/source/examples/unresolved-cfd-dem/liquid-solid-fluidized-bed/liquid-solid-fluidized-bed.rst

      set ilu preconditioner relative tolerance = 1.00
      set verbosity                             = verbose
    end

------------------------------
Running the CFD-DEM Simulation
------------------------------

The simulation is run (on 8 core) using the ``cfd_dem_coupling`` application as follows:

.. code-block:: text

    mpirun -np 8 cfd_dem_coupling liquid-solid-fluidized-bed.prm

--------
Results
--------

We briefly comment on some results that can be extracted from this example.

.. important::

    This example includes a post-processing file written in Python and using `lethe_pyvista_tools <../../dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.html>`_. module.

    To use the code, run ``python3 lsfb_postprocessing.py $PATH_TO_YOUR_CASE_FOLDER``. The code will generate several graphics showing the pressure profile within the bed, which are going to be stored in ``$PATH_TO_YOUR_CASE_FOLDER/P_x``. It will also generate a ``deltaP_t.csv`` file with the total pressure difference for each time-step. Additionally, it generates a void fraction as a function of time graphic (``eps_t.png``).

> Side view

Reviewer: ```suggestion
Side View
```

~~~~~~~~~~~

Here we show comparison between the experimentally observed and simulated behavior of the liquid-solid fluidized bed with alumina.

The void fraction and velocity profile of the fluid are also shown.

.. raw:: html

    <p align="center"><iframe width="560" height="315" src="https://www.youtube.com/embed/Ra7d-p7wD8Y" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

> Total pressure drop and bed expansion

Reviewer: ```suggestion
Total Pressure Drop and Bed Expansion
```

where :math:`H` is the total bed height, :math:`\bar{\varepsilon}_f` is the average fluid fraction (void fraction) at the bed region, :math:`\rho_p` and :math:`\rho_f` are the densities of the particles and the fluid (respectively), and :math:`A` is the cross-section area of the equipment.

Liquid fluidized beds are very uniform in terms of particles distribution, resulting in an uniform distribution of  :math:`\varepsilon_f` along the be height. From this hypothesis, we can conclude that, for a constant and uniform fluid inlet flow rate, the pressure slope is:

.. math::

    \left.- \frac{\mathrm{d} p }{\mathrm{d} z}\right|_{z = 0}^{z = H}  \approx \mathrm{constant}

With the pressure slope, it is also possible to determine the bed void fraction manipulating the first equation, which gives:

.. math::

    \bar{\varepsilon}_f = 1 - \frac{\left.- \frac{\mathrm{d} p }{\mathrm{d} z}\right|_{z = 0}^{z = H} }{(\rho_p - \rho_f)g}

The resulting behavior of the pressure along the bed height and the void fraction with time is shown in the following animation.

.. image:: images/pressure_time.gif
    :alt: Pressure drop as a function of time
    :align: center
    :name: press_t

> Particles dynamics

Reviewer: ```suggestion
Particles Dynamics
```

~~~~~~~~~~~~~~~~~~~~

Since the fluidization occurs in a high density fluid, the density difference between alginate and alumina particles have a significant impact on the velocity of the particles inside the bed.

The following animation is in real time. It is possible to notice that, for a similar bed height, the bed of alumina particles expands way faster than the alginate.

.. raw:: html

    <p align="center"><iframe width="560" height="315" src="https://www.youtube.com/embed/kMp86PdZ6tU" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" allowfullscreen></iframe>

> Notes on performance

Reviewer: ```suggestion
Notes on Performance
```

==================================
Liquid-Solid Fluidized Bed
==================================

It is strongly recommended to visit `DEM parameters <../../../parameters/dem/dem.html>`_  and `CFD-DEM parameters <../../../parameters/unresolved-cfd-dem/unresolved-cfd-dem.html>`_ for more detailed information on the concepts and physical meaning of the parameters ind DEM and CFD-DEM.

----------------------------------
Features
----------------------------------
- Solvers: ``dem`` and ``cfd_dem_coupling``
- Three-dimensional problem
- Displays the selection of models and physical properties.
- Simulates a solid-liquid fluidized bed.
- Post-processing code available.

---------------------------
Files Used in This Example
---------------------------
``/examples/unresolved-cfd-dem/gas-solid-fluidized-bed/liquid-solid-fluidized-bed.prm``
``/examples/unresolved-cfd-dem/gas-solid-fluidized-bed/dem-packing-in-lsfb.prm``

> ``/examples/unresolved-cfd-dem/gas-solid-fluidized-bed/lsfb_postprocessing.py``

Reviewer 2: Please do so

.. figure:: images/experimental_setup.png
    :alt: particle packing
    :align: center

    Pilot-scale fluidized bed used as reference for simulations in this example. Image includes (a) schematic representation of the bed and (b) picture of the equipment in oparation. Adapted from Ferreira *et al* [1].

-------------------
DEM Parameter File
-------------------

As in the other examples of this documentation, we use Lethe-DEM to fill the bed with particles. We enable check-pointing in order to write the DEM checkpoint files which will be used as the starting point of the CFD-DEM simulation. Then, we use the ``cfd_dem_coupling`` solver within Lethe to simulate the fluidization of the particles by initially reading the checkpoint files from the DEM simulation.

All parameter subsections are described in the `parameter section <../../../parameters/parameters.html>`_ of the documentation.

To set-up the cylinder fluidized bed case, we first fill the bed with particles.

We first introduce the different sections of the parameter file ``dem-packing-in-lsfb.prm`` needed to run this simulation.

> The parameter is divided into `subsections`, presented individually as follows.

Reviewer 2: Just removed

Mesh
~~~~~

In this example, we are simulating a cylindrical fluidized bed that has a half length of 0.55 m (10 cm higher than the fluidization region in the experimental setup), and a diameter of 10 cm. We use the `subdivided_cylinder GridGenerator <https://www.dealii.org/current/doxygen/deal.II/namespaceGridGenerator.html#a95f6e6a7ae2fe3a862df035dd2cb4467:~:text=%E2%97%86-,subdivided_cylinder,-()>`_  in order to generate the mesh. The square bed is divided 132 times in the x direction (height) and 12 times in y and z directions (6 times in the radius). The following portion of the DEM parameter file shows the function called:

.. code-block:: text

    subsection mesh
      set type               = dealii
      set grid type          = subdivided_cylinder
      set grid arguments     = 33:0.05:0.55
      set initial refinement = 2
      set expand particle-wall contact search = true
    end

.. note::
    Note that, since the mesh is cylindrical, ``set expand particle-wall contact search = true``. Details in the `DEM mesh parameters guide <../../../parameters/dem/mesh.html>`_.

> A cross section of the resulting mesh is presented in the following figure.

Reviewer 2: ```suggestion
A cross-section of the resulting mesh is presented in the following figure.
```

Floating Walls
~~~~~~~~~~~~~~~~~~~

A floating wall is added 10 cm above the bottom of the mesh, so that void fraction discontinuities can be avoided. The remaining region above the floating wall is 1 m high, as in the experimental setup.

.. code-block:: text

    subsection floating walls
      set number of floating walls = 1
      subsection wall 0
        subsection point on wall
          set x = -0.45
          set y = 0
          set z = 0
        end
      subsection normal vector
        set nx = 1
        set ny = 0
        set nz = 0
      end
      set start time = 0
      set end time   = 50
      end
    end

.. note::
    Note that ``end time`` is higher than ``time end`` in ``simulation control``, so that the floating wall does not disappear in the middle of the simulation.

Simulation Control
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Here, we define the time-step and the simulation end time.

.. code-block:: text

    subsection simulation control
      set time step        = 0.000005
      set time end         = 10.5
      set log frequency    = 1000
      set output frequency = 20000
      set output path      = ./output_dem/
    end

> It is important to define ``time end`` according to particles insertion (explained further in this example).

Reviewer 2: This is very vague. I would say this it is important to define the time end to include also the time required to insert the particles in the first simulation

---------------------------
Running the DEM Simulation
---------------------------
Launching the simulation is as simple as specifying the executable name and the parameter file. Assuming that the ``dem`` executable is within your path, the simulation can be launched on a single processor by typing:

.. code-block:: text

  dem dem-packing-in-fluidized-bed.prm

or in parallel (where 8 represents the number of processors)

.. code-block:: text

  mpirun -np 8 dem dem-packing-in-fluidized-bed.prm

Lethe will generate a number of files. The most important one bears the extension ``.pvd``. It can be read by popular visualization programs such as `Paraview <https://www.paraview.org/>`_. 

.. note:: 
    Running the packing of alumina particles should take approximately 4 hours and 15 minutes on 16 cores. For the alginate particles, it takes approximately 5 hours and 52 minutes (``time end = 9.9``).

> After the particles have been packed inside the cylinder, it is now possible to simulate the fluidization of particles.

Reviewer 3: ```suggestion
Now that the particles have been packed inside the cylinder, it is possible to simulate the fluidization of particles.
```

> After the particles have been packed inside the cylinder, it is now possible to simulate the fluidization of particles.

Reviewer 2: Do you really need 10 seconds to pack the cylinder? Seems like a lot!
Are you sure it takes this much time?

Author: The insertion parameters are very outdated. I just optimized it. I am waiting on Narval, but 2.5 s should be enough with the new parameters.

.. code-block:: text

    subsection physical properties
      subsection fluid 0
        set kinematic viscosity = 0.0000008379
        set density             = 997
      end
    end

Initial Conditions
~~~~~~~~~~~~~~~~~~

For the initial conditions, we choose zero initial conditions for the velocity. 

.. code-block:: text

    subsection initial conditions
      set type = nodal
      subsection uvwp
          set Function expression = 0; 0; 0; 0
      end
    end
 

Boundary Conditions
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

For the boundary conditions, we choose a slip boundary condition on the walls (ID = 0) and an inlet velocity of 0.157033 m/s at the lower face of the bed (ID = 1).

.. code-block:: text

    subsection boundary conditions
      set number = 2
      subsection bc 0
        set id   = 0
        set type = slip
      end
      subsection bc 1
        set id   = 1
        set type = function
        subsection u
          set Function expression = 0.157033
        end
        subsection v
          set Function expression = 0
        end
        subsection w
          set Function expression = 0
        end
      end
    end

> The additional sections for the CFD-DEM simulations are the void fraction subsection and the CFD-DEM subsection. These subsections are described in detail in the `CFD-DEM parameters <../../../parameters/unresolved-cfd-dem/unresolved-cfd-dem.html>`_ .

Reviewer 3: ```suggestion
The following sections for the CFD-DEM simulations are the void fraction subsection and the CFD-DEM subsection. These subsections are described in detail in the `CFD-DEM parameters <../../../parameters/unresolved-cfd-dem/unresolved-cfd-dem.html>`_ .
```

Void Fraction
~~~~~~~~~~~~~~~

We choose the `particle centroid method (PCM) <../../../parameters/unresolved-cfd-dem/void-fraction.html>`_ to calculate void fraction. The ``l2 smoothing factor`` we choose is around the square of twice the particle’s diameter, as in the other examples.
 
.. code-block:: text

    subsection void fraction
      set mode                = pcm
      set read dem            = true
      set dem file name       = dem
      set l2 smoothing factor = 0.000028387584
      set bound void fraction = false
    end

.. note::

>     Note that void fraction is not bound in this case. The particles used in this example obligates us to use a very coarse mesh. Bounding void fraction lead to instability in the present case.

Reviewer 3: ```suggestion
    Note that void fraction is not bound in this case. The size particles used in this example obligates us to use a very coarse mesh. Bounding void fraction would lead to instability in the present case.
```

CFD-DEM
~~~~~~~~~~

> Different from gas-solid fluidized beds, buoyancy, pressure force, shear stress are not negligible. All these forces are considered in this example.

Reviewer 3: ```suggestion
In this case, a difference from gas-solid fluidized beds is that buoyancy, pressure force and shear stress are not negligible. All these forces are considered in this example.
```

      set drag model         = rong
      set buoyancy force     = true
      set shear force        = true
      set pressure force     = true
      set saffman lift force = true
      set coupling frequency = 100
      set void fraction time derivative = false
    end

.. warning::
    Void-fraction time-derivative lead to significant instability in the case of liquid-fluidized beds, hence we do not use it.

Non-linear Solver
~~~~~~~~~~~~~~~~~

We use the inexact Newton non-linear solver to minimize the number of time the matrix of the system is assembled. This is used to increase the speed of the simulation, since the matrix assembly requires significant computations.

.. code-block:: text

  subsection non-linear solver
    set solver           = inexact_newton
    set tolerance        = 1e-10
    set max iterations   = 10
    set verbosity        = verbose
  end

Linear Solver
~~~~~~~~~~~~~

.. code-block:: text

    subsection linear solver
      set method                                = gmres
      set max iters                             = 5000
      set relative residual                     = 1e-3
      set minimum residual                      = 1e-11

>       set ilu preconditioner fill               = 1

Reviewer 2: Do you really need a fill level of 1?

Author: Nope. I am testing for performance, but it does seem necessary.

Author: Performance-wise it was terrible compared to fill level 1. I tested on 32 cores and it took about 3 times as much.

      set ilu preconditioner absolute tolerance = 1e-14
      set ilu preconditioner relative tolerance = 1.00
      set verbosity                             = verbose
    end

------------------------------
Running the CFD-DEM Simulation
------------------------------

> The simulation is run (on 8 core) using the ``cfd_dem_coupling`` application as following:

Reviewer 3: ```suggestion
The simulation is run (on 8 core) using the ``cfd_dem_coupling`` application as follows:
```

.. code-block:: text

    mpirun -np 8 cfd_dem_coupling dem-packing-in-fluidized-bed.prm

.. note::
    Running the packing of alumina particles should take approximately 4 hours and 15 minutes on 16 cores.

--------
Results
--------

We briefly comment on some results that can be extracted from this example.

.. important::

>     This example includes a post-processing file written in Python and using `lethe_pyvista_tools <../../dem/rotating-drum/small-scale-rotating-drum-post-processing.html>`_. module.

Reviewer 3: ```suggestion
    This example includes a post-processing file written in Python and using `lethe_pyvista_tools <../../dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.html>`_. module.
```

>     To use the code, run ``python3 lsfb_postprocessing.py $PATH_TO_YOUR_CASE_FOLDER``. The code will generate several graphics showing the pressure profile within the bed, which are going to be stored in ``$PATH_TO_YOUR_CASE_FOLDER/P_x``. It will also generate a ``deltaP_t.csv`` files with the total pressure difference for each time-step. Additionally, it generates a void fraction as a function of time graphic (``eps_t.png``).

Reviewer 3: ```suggestion
    To use the code, run ``python3 lsfb_postprocessing.py $PATH_TO_YOUR_CASE_FOLDER``. The code will generate several graphics showing the pressure profile within the bed, which are going to be stored in ``$PATH_TO_YOUR_CASE_FOLDER/P_x``. It will also generate a ``deltaP_t.csv`` file with the total pressure difference for each time-step. Additionally, it generates a void fraction as a function of time graphic (``eps_t.png``).
```

.. code-block:: text

    subsection boundary conditions
      set number = 2
      subsection bc 0
        set id   = 0
        set type = slip
      end
      subsection bc 1
        set id   = 1
        set type = function
        subsection u
          set Function expression = 0.157033
        end
        subsection v
          set Function expression = 0
        end
        subsection w
          set Function expression = 0
        end
      end
    end

The following sections for the CFD-DEM simulations are the void fraction subsection and the CFD-DEM subsection. These subsections are described in detail in the `CFD-DEM parameters <../../../parameters/unresolved-cfd-dem/unresolved-cfd-dem.html>`_ .

Void Fraction
~~~~~~~~~~~~~~~

We choose the `particle centroid method (PCM) <../../../parameters/unresolved-cfd-dem/void-fraction.html>`_ to calculate void fraction. The ``l2 smoothing factor`` we choose is around the square of twice the particle’s diameter, as in the other examples.
 
.. code-block:: text

    subsection void fraction
      set mode                = pcm
      set read dem            = true
      set dem file name       = dem

>       set l2 smoothing factor = 0.000028387584

Reviewer 2: ```suggestion
      set l2 smoothing factor = 2.8387584e-5
```
Better to use scientific formatting when you have so many zeros

>       set bound void fraction = false

Reviewer 2: Is it not set to false by default anyway?

.. code-block:: text

  subsection non-linear solver
    set solver           = inexact_newton
    set tolerance        = 1e-10
    set max iterations   = 10
    set verbosity        = verbose
  end

Linear Solver
~~~~~~~~~~~~~

.. code-block:: text

    subsection linear solver
      set method                                = gmres
      set max iters                             = 5000
      set relative residual                     = 1e-3
      set minimum residual                      = 1e-11
      set ilu preconditioner fill               = 1
      set ilu preconditioner absolute tolerance = 1e-14
      set ilu preconditioner relative tolerance = 1.00
      set verbosity                             = verbose
    end

------------------------------
Running the CFD-DEM Simulation
------------------------------

The simulation is run (on :math:`8 \: \text{cores}`) using the ``cfd_dem_coupling`` application as follows:

.. code-block:: text
  :class: copy-button

>     mpirun -np 8 cfd_dem_coupling liquid-solid-fluidized-bed.prm

Reviewer 2: What's the new simulation time here?
